### ISSUE TITLE
Mappings in qt5_4.imp not working

### ISSUE THREAD

**Issue by estan** [include-what-you-use/include-what-you-use#335]
With

```cpp
#include <QDebug>

int main() {
    qWarning() << "sweet";
}
```

I get

```
main.cpp should add these lines:
#include <qlogging.h>      // for qWarning
#include <QMessageLogger>  // for QMessageLogger

main.cpp should remove these lines:

The full include-list for main.cpp:
#include <qlogging.h>      // for qWarning
#include <QDebug>          // for QDebug
#include <QMessageLogger>  // for QMessageLogger
```

With more complex Qt code, I also get many other wrong suggestions (`qobjectdefs.h`, `qnamespace.h`, ...) that, by the looks of it, the mapping file should take care of.

This is with Qt 5.7.0. I know the mapping file is made for Qt 5.4, but it should still catch these things, right?

**Comment by estan** [comment]
[iwyu.out.txt](https://github.com/include-what-you-use/include-what-you-use/files/342221/iwyu.out.txt) is the output of the invocation

```
include-what-you-use -Xiwyu --verbose=5 -Xiwyu --mapping_file=qt5_4.imp -std=c++11 -c -pipe -O2 -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -I. -I. -isystem /usr/include/qt -isystem /usr/include/qt/QtGui -isystem /usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o main.o main.cpp
```

which is hopefully of some use to debug the issue.

**Comment by estan** [comment]
This is with `include-what-you-use 0.6 based on clang version 3.8.0 (tags/RELEASE_380/final)` BTW.

**Comment by ThosRTanner** [comment]
That's odd. As far as I understand it, this line:

     { include: [ "@\"(QtCore/)?qlogging\\.h\"", "private", "<QtDebug>", "public" ] }, #qDebug, qWarning, etc

should remove qlogging.h and ask for <QtDebug> (not QDebug. I plumped for the former, because there's two headers, one of which gets a more flexible version of qDebug()). Having said which, I have no idea where QMessageLogger is coming from. I suppose the way the headers are included might have changed so things are no longer recognised. I should get sufficient courage up to upgrade my QT install soon and will have a look see.

**Comment by estan** [comment]
Yep, I thought it strange as well after looking at your rules. No hurry, I can live without IWYU for this project for now. Tell me if you want me to try/investigate something.

**Comment by estan** [comment]
Hum, as I thought the `\"`s inside the regexp were strange (but didn't know if it was something special with the mapping file syntax I wasn't aware of), I tried changing the rule from

    { include: [ "@\"(QtCore/)?qlogging\\.h\"", "private", "<QtDebug>", "public" ] }, #qDebug, qWarning, etc

to

    { include: [ "@<(QtCore/)?qlogging\\.h>", "private", "<QtDebug>", "public" ] }, #qDebug, qWarning, etc

And it seems to have fixed things:

```
main.cpp should add these lines:
#include <QMessageLogger>  // for QMessageLogger
#include <QtDebug>         // for qWarning

main.cpp should remove these lines:

The full include-list for main.cpp:
#include <QDebug>          // for QDebug
#include <QMessageLogger>  // for QMessageLogger
#include <QtDebug>         // for qWarning
```

It now reports `QtDebug` (I didn't change anything else).

What is the reason for the `\"`s in the rules?

**Comment by estan** [comment]
I think that's it. I think the rules should use `<` and `>`, since that's how you normally include Qt headers?

**Comment by estan** [comment]
If I do the replacements on all the rules, it seems to work fine on my non-trivial project.

One thing that is a little annoying though is that it suggests I use

```cpp
#include <QMetaObject>      // for QMetaObject
#include <QObject>          // for Q_OBJECT
```

whenever I declare a QObject-derived class, e.g like this

```cpp
#ifndef MAINWINDOW_H
#define MAINWINDOW_H

#include "ui_MainWindow.h"

#include <QObject>
#include <QMetaObject>
#include <QMainWindow>

class MainWindow : public QMainWindow, private Ui::MainWindow
{
	Q_OBJECT

public:
	explicit MainWindow(QWidget *parent = 0);
};

#endif // MAINWINDOW_H
```

Including those two are normally not done in Qt coding, where you instead rely on getting them transitively from the header of the QObject class you are deriving from, but I guess it's hard to make a rule to avoid it reporting that (the QMetaObject is added by `moc`, that's why it's reported, so it's logical at least).

**Comment by estan** [comment]
Actually, even with fixing the `\#` vs `<` issues. I get

    #include <QMessageLogger>       // for QMessageLogger

But I think that has to do with `qWarning` being a macro that is expanded to use `QMessageLogger` (haven't checked), so I guess that's hard to fix as well.

**Comment by ThosRTanner** [comment]
The < and " depends on what qt does internally. On my windows box on qt5.4 it uses "" all over the place.

**Comment by estan** [comment]
Aha, then I misunderstood how the rules work. I thought the left hand side of a rule was matched against whatever include-what-you-get would normally suggest? That suggestion would always be with `<` and `>`, wouldn't it?

**Comment by ThosRTanner** [comment]
for some reason i thought it always came up with "s when checking for filename matches like that. I'd have to look closer.

I don't know if it's to do with using -isystem. My build command is rather less subtle and  uses -I all over and then has `--system-header-prefix=Q`. I had a poke round my QT headers and the all use #include <> internally, so it's probably not the </" that's causing the .imp file not to work.

In the 5.5, QMessageLogger is a friend class. Assuming that's the same in 5.7, I can't see why it'd be being shown as a needed include, unless something in IWYU has changed.

**Comment by estan** [comment]
Alright, all I know is that here `<` and `>` works, and double quotes doesn't. I don't understand why it would depend on what Qt uses internally. I would think it depends on what include-what-you-get uses when it passes a header that must be included to the mapping code. But maybe I misunderstand how the tool works.

Regarding `QMessageLogger`, at least here, `qWarning` is a macro, so my call to `qWarning` expands to:

    QMessageLogger(nullptr, 0, nullptr).warning("sweet");

**Comment by estan** [comment]
I get the same with `-I` instead of `-isystem` BTW:

```
[estan@pyret iwyutest]$ include-what-you-use -std=c++11 -c -pipe -O2 -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -Wall -W -D_REENTRANT -fPIC -DQT_NO_DEBUG -DQT_GUI_LIB -DQT_CORE_LIB -I. -I. -I/usr/include/qt -I/usr/include/qt/QtGui -I/usr/include/qt/QtCore -I. -I/usr/lib/qt/mkspecs/linux-g++ -o main.o main.cpp

main.cpp should add these lines:
#include "qdebug.h"    // for QDebug
#include "qlogging.h"  // for QMessageLogger, qWarning

main.cpp should remove these lines:
- #include <QDebug>  // lines 1-1

The full include-list for main.cpp:
#include "qdebug.h"    // for QDebug
#include "qlogging.h"  // for QMessageLogger, qWarning
---
[estan@pyret iwyutest]$
```

The compile line I used is the one `CMake` 3.5 generates for a Qt project on Linux.

**Comment by estan** [comment]
Interestingly, I don't get the problem with `qWarning` -> `QMessageLogger` at work, where I run Qt 5.5.1.

**Comment by ThosRTanner** [comment]
I have tried building with 5.7 and it seems to work fine as is. I get the unexpected QMessageLogger but all my output is like this:

```
C:\Apps\include-what-you-use\include-what-you-use-0.6-x86-win32\include-what-you-use\bin\include-what-you-use.exe -Xiwyu --transitive_includes_only -D_MT -D_DLL -m32 -DBOOST_USE_WINDOWS_H -Wno-inconsistent-missing-override --system-header-prefix=Q --system-header-prefix=boost/ -fmsc-version=1900 -fexceptions -fmsc-version=1900 -D_DEBUG -Xiwyu --mapping_file=C:\Apps\include-what-you-use\include-what-you-use-0.6-x86-win32\include-what-you-use\share\qt5_4.imp -Xiwyu --mapping_file=modorganizer\win.imp -Xiwyu --mapping_file=modorganizer\mappings.imp -DUNICODE -DWIN32 -DNOMINMAX -DQT_CORE_LIB -D_UNICODE -D_CRT_SECURE_NO_WARNINGS -IC:\Apps\Qt\5.7\msvc2013\mkspecs\win32-msvc2013 -IC:\Apps\Qt\5.7\msvc2013\include -IC:\Apps\Qt\5.7\msvc2013\include -IC:\Apps\Qt\5.7\msvc2013\include\QTCore source\helper\main.cpp

source/helper/main.cpp should add these lines:
#include <stdio.h>         // for getchar
#include <string.h>        // for wcscmp
#include <QByteArray>      // for QByteArray
#include <QFileInfo>       // for QFileInfo
#include <QMessageLogger>  // for QMessageLogger
#include <QString>         // for QString
#include <QStringList>     // for QStringList
#include <QtDebug>         // for qCritical, qDebug
#include <QtGlobal>        // for QT_VERSION, QT_VERSION_CHECK
#include <string>          // for wstring

source/helper/main.cpp should remove these lines:
- #include <QSettings>  // lines 23-23
- #include <QtCore/QCoreApplication>  // lines 22-22

The full include-list for source/helper/main.cpp:
#include <Windows.h>       // for GetLastError, FILETIME, CloseHandle, Creat...
#include <stdio.h>         // for getchar
#include <string.h>        // for wcscmp
#include <QByteArray>      // for QByteArray
#include <QDir>            // for QDir, QDir::Filter::Files, QDir::Filters
#include <QDirIterator>    // for QDirIterator
#include <QFileInfo>       // for QFileInfo
#include <QMessageLogger>  // for QMessageLogger
#include <QString>         // for QString
#include <QStringList>     // for QStringList
#include <QtDebug>         // for qCritical, qDebug
#include <QtGlobal>        // for QT_VERSION, QT_VERSION_CHECK
#include <string>          // for wstring
#include "privileges.h"    // for SetOwner
---
```

So the `<>`s are all there as expected (this is a windows build, not unix, I don't know if that makes a difference), and I need to persuade IWYU to treat some files as being exported by others (see issue referencing this, as I'm not sure I can)

**Comment by estan** [comment]
Hm okay. Then there definitely must be some difference in our environments, because I have to change all the rules to use `<` and `>` instead of `\"`, otherwise none of them match.

**Comment by ThosRTanner** [comment]
The `QMessageLogger` thing is definitely a different problem. I think `qDebug() << `and `qDebug("format") ` are defined in different headers and this happens with QT5.5 as well - I think the `<QMessageLogger>` header doesn't exist and it's being mapped wrong.

**Comment by ThosRTanner** [comment]
I had a little fiddle, and firstly, the QT Documentation seems to be misleading. It says use `QtDebug` for the `operator<<` version of `qWarning`, etc. But that's in `<QDebug>` (which both appear to include `"qdebug.h"`). Not sure how it's getting into QMessageLogger. might be time for a high level of verbosity again

**Comment by ThosRTanner** [comment]
OK, QMessageLogger is coming out because of something that changed between 0.4 and 0.6. It's not there with 0.4

**Comment by ThosRTanner** [comment]
It's not there on 0.5 either. So this is a change in 0.6

**Comment by kimgr** [comment]
It could be something to do with the code I added for attributing uses to macros vs macro instantiations. Is there a macro involved there somewhere?

**Comment by kimgr** [comment]
> As to the < and " problem, I guess we'll have to await a comment from the experts

I'm not entirely sure why the quotes are part of the mappings, it's a historical artifact from before I joined project. It works for standard library headers (where you're required to use `<>`), but not so well for third-parties where you can use either `<>` or `""` interchangeably.

Maybe it's time to try a patch where we remove all quoting from mappings...

**Comment by estan** [comment]
> It could be something to do with the code I added for attributing uses to macros vs macro instantiations. Is there a macro involved there somewhere?

Yep, `qDebug`, `qInfo`, `qWarning`, `qCritical` and `qFatal` are all macros defined as

```cpp
#define qDebug QMessageLogger(QT_MESSAGELOG_FILE, QT_MESSAGELOG_LINE, QT_MESSAGELOG_FUNC).debug
#define qInfo QMessageLogger(QT_MESSAGELOG_FILE, QT_MESSAGELOG_LINE, QT_MESSAGELOG_FUNC).info
#define qWarning QMessageLogger(QT_MESSAGELOG_FILE, QT_MESSAGELOG_LINE, QT_MESSAGELOG_FUNC).warning
#define qCritical QMessageLogger(QT_MESSAGELOG_FILE, QT_MESSAGELOG_LINE, QT_MESSAGELOG_FUNC).critical
#define qFatal QMessageLogger(QT_MESSAGELOG_FILE, QT_MESSAGELOG_LINE, QT_MESSAGELOG_FUNC).fatal
```

**Comment by estan** [comment]
> Maybe it's time to try a patch where we remove all quoting from mappings...

I think that would make sense. There's apparently something that makes mine and @ThosRTanner's systems different: I have to use `<` and `>` for the rules to match, and he has to use `\"`. To me it makes more sense that `<` have to be used, since the Qt includes are normally in paths specified with `-I`, for which it's customary to use `<` and `>`, not in some relative directory to the project source.

Then of course is the question of what the tool should use in the output when it suggests includes. I think `<` which is used now is okay (because that's always used in the output, or?), unless the tool is smart enough to realize which should use `"`.

**Comment by ThosRTanner** [comment]
could iwyu look to see where it found the header file and use " or < as appropriate? (i.e. use the -isystem stuff)

**Comment by estan** [comment]
To clarify: I still agree with @kimgr that the rules should not have to contain `<` / `>` or `\"` at all, but should simply be matched against the header file names. I just thought that in the case of Qt, which is normally used as a "public" library with `<` / `>`, it makes more sense that the rules required `<` / `>`. But as it is right now, it seems it can be system-dependent what the rules should use, which is not good.

**Comment by ThosRTanner** [comment]
I tried defining qDebug &c as symbols that should be found in <QtDebug> but that didn't help much

**Comment by kimgr** [comment]
> could iwyu look to see where it found the header file and use " or < as appropriate? (i.e. use
> the -isystem stuff)

It does, so if you specify -isystem for everything you want `<>` for and -I for everything else, things should work.

Problem is people usually don't use `-isystem`, so this tends to be a workaround that doesn't survive for very long.

**Comment by nocnokneo** [comment]
I'm also being hit by this issue. What's the status of the ideal to remove all quoting from mappings? That approach makes sense to me.

**Comment by kimgr** [comment]
We haven't come around to it yet. I'd like to finalize the macro handling before moving onto anything else, but we're still struggling with Clang-induced test failures on master.

**Comment by firewave** [comment]
Qt 5.4 is long EOL, so I think this can be closed.

The extremely old Qt mappings which are part of the repo should also be dropped.
